deps/thpool/thpool.c

```diff
@@ -209,180 +236,212 @@ int redisearch_thpool_add_work(redisearch_thpool_t* thpool_p, void (*function_p)
-  time_spec.tv_sec = tp.tv_sec;
-  time_spec.tv_nsec = tp.tv_usec * 1000;
-  time_spec.tv_nsec += timeout * 1000000;
-  // Check if the nanoseconds field has a carry-out (it cannot hold a value which is lager
-  // than 1 sec).
-  if (time_spec.tv_nsec >= 1000000000) {
-    time_spec.tv_sec++;
-    time_spec.tv_nsec -= 1000000000;
-  }
+static void redisearch_thpool_push_chain_init_threads(
+    redisearch_thpool_t *thpool_p, struct job *f_newjob_p,
+    struct job *l_newjob_p, size_t n, thpool_priority priority) {
+  redisearch_thpool_lock(thpool_p);
+  priority_queue_push_chain_unsafe(&thpool_p->jobqueues, f_newjob_p, l_newjob_p,
+                                   n, priority);
+  redisearch_thpool_unlock(thpool_p);
+
+  /* Initialize threads if needed */
+  redisearch_thpool_verify_init(thpool_p);
+}
 
-  pthread_mutex_lock(&thpool_p->thcount_lock);
-  while (priority_queue_len(&thpool_p->jobqueue) > threshold) {
-    int rc = pthread_cond_timedwait(&thpool_p->threads_all_idle, &thpool_p->thcount_lock, &time_spec);
-    if (rc == ETIMEDOUT) {
-      pthread_mutex_unlock(&thpool_p->thcount_lock);
+/* Wait until all jobs have finished */
+void redisearch_thpool_wait(redisearch_thpool_t *thpool_p) {
+  priority_queue *priority_queue_p = &thpool_p->jobqueues;
+  long msec_timeout = 100;
+  redisearch_thpool_drain(thpool_p, msec_timeout, NULL, NULL, 0);
+}
+
+void redisearch_thpool_drain(redisearch_thpool_t *thpool_p, long timeout,
+                             yieldFunc yieldCB, void *yield_ctx,
+                             size_t threshold) {
+  long usec_timeout = 1000 * timeout;
+  while (priority_queue_unprocessed_jobs(&thpool_p->jobqueues) > threshold) {
+    usleep(usec_timeout);
+    if (yieldCB)
       yieldCB(yield_ctx);
-      gettimeofday(&tp, NULL);
-      time_spec.tv_sec = tp.tv_sec;
-      time_spec.tv_nsec = tp.tv_usec * 1000;
-      time_spec.tv_nsec += timeout * 1000000;
-      // Check if the nanoseconds field has a carry-out (it cannot hold a value which is lager
-      // than 1 sec).
-      if (time_spec.tv_nsec >= 1000000000) {
-        time_spec.tv_sec++;
-        time_spec.tv_nsec -= 1000000000;
-      }
-      pthread_mutex_lock(&thpool_p->thcount_lock);
-    } else {
-      assert(rc == 0);
-    }
   }
-  pthread_mutex_unlock(&thpool_p->thcount_lock);
 }
 
-static void redisearch_thpool_terminate_threads_common(redisearch_thpool_t* thpool_p, thpool_state new_state) {
+void redisearch_thpool_terminate_reset_threads(redisearch_thpool_t *thpool_p) {
   RedisModule_Assert(thpool_p);
+  if (!priority_queue_is_empty(&thpool_p->jobqueues)) {
+    LOG_IF_EXISTS(
+        "warning",
+        "Terminate threadpool's thread was called when the jobq is not empty");
+  }
+  // Threads might be in terminate when empty state, we must lock before we read
+  // `num_threads_alive` to ensure they don't die (i.e check that jobq is not
+  // empty) to read `num_threads_alive`
+  redisearch_thpool_lock(thpool_p);
```

Comment:
**There should not be any jobs in progress when calling this function.** Note that the scenario you described is pretty optimistic because if we called this function when the jobq is not empty, it **might not** get empty by the time we finish all the threads and jobs will remain undone, and this is a situation we want to avoid.
If the situation you described occurred, then someone misused the thpool. 
Hence, it doesn't matter if we release the lock

Reply:
I agree that logically there is no problem, I meant that it Is better practice to lock once if possible